### ISSUE TITLE
BVH.shapecast's intersectsBounds is never ran with leafNodes

### ISSUE THREAD

**Issue by codechriscode** [gkjohnson/three-mesh-bvh#530]
### **Describe the bug**

I'm using `BVH.shapecast({intersectsBounds, intersectsTriangles})` to select vertices and triangles whose bounding boxes `min` or `max` are contained in a `three.Sphere`. 
I logged the arguments of all `intersectsBounds` function calls and 
- for meshes which work, `intersectsBounds` is ran with bounding boxes of depth 10+, finiding leaf nodes.
- for meshes which break, `intersectsBounds` is only ever ran with bounding boxes of depth 1 or 2, never leaf nodes. They are too big and are never contained in the `three.Sphere`.

The meshes result of OBJ files. MeshBVHVisualizer shows BVs as expected, but I'm not sure the hierarchy tree is correct since depth1 BVs are not sending me forward to leaf nodes. Is there a way to visualize the hierarchy to see if the tree data structure is well formed?

### **Screenshots and expected behavior**
Logs recorded every time `bvh.shapecast` calls `intersectsBounds`.

**Expected Behaviour**
- `bvh.shapecast` offers all intersecting `box` along the hierarchy tree, often small leaf nodes of depth > 10
![image](https://github.com/gkjohnson/three-mesh-bvh/assets/58148387/5ea834be-4d82-4199-94c6-9c92cbb8249e)

**Wrong Behavior**
- `bvh.shapecast` only ever offers `box` at the top of the hierarchy tree, big top nodes of depth 1 or 2
![image](https://github.com/gkjohnson/three-mesh-bvh/assets/58148387/4b810166-dd46-481b-99b9-2a00813b8813)

### **Code**

```js
boundingVolumeHierarchy.shapecast({
  intersectsBounds: (box, isLeaf, _, depth, nodeIndex) => {
    console.log(box, isLeaf, depth);
    // tempVec is either a copy of box.min or box.max
    // sphere has radius 4 and the position coords of the click/location, which are all correct
    console.log(tempVec, sphere);
  }
})
```

### **Platform:**
 - Device: [Desktop]
 - OS: [Windows, MacOS, Linux]
 - Browser: [Chrome, Firefox, Safari, Edge]
 - Three.js version: [0.138.0]
 - Library version: [0.5.5]

**Comment by gkjohnson** [comment]
I'm sorry I'm having a hard to following what the problem is. Can you provide a jsfiddle showing what the issue is and what the expected results are?

**Comment by codechriscode** [comment]
I didn't put it in a jsfiddle because there's some big BufferGeometry from specific OBJ files that are private. I'll try to partly reproduce it.

Looking at the logs, what's not working is that the `spheres (x,y,z,radius)` are intersecting with upper bounding volumes (depth = 1, or depth = 2) but never with volumes down the hierarchy.
That means when `bvh.shapecast` is called and runs the `intersectsBounds` function, it is never passing a (box) argument beyond the first or second bounding volumes.

As you'd expect, depth =1 bouding volumes are too big to be contained in a sphere. `bvh.shapecast` is never offering smaller, specific bounding volume down the hierarchy chain for me to intercept.
You can compare the logs of expected behavior and broken behavior and see the box depth and size at each `intersectsBounds` call from `bvh.shapecast`.

**Comment by gkjohnson** [comment]
It's still not clear how you're using the function and what you're returning from it. If you're just running this snippet then the return value of "undefined" is not valid. At best it gets interpreted as NOT_INTERSECTED resulting in the traversal ending:

```js
boundingVolumeHierarchy.shapecast({
  intersectsBounds: (box, isLeaf, _, depth, nodeIndex) => {
    console.log(box, isLeaf, depth);
    // tempVec is either a copy of box.min or box.max
    // sphere has radius 4 and the position coords of the click/location, which are all correct
    console.log(tempVec, sphere);
  }
})
```

If you're doing something else then unfortunately I don't have the bandwidth to guess at what your implementation currently is or try to reproduce results of an undemonstrated bug. I need sample code to understand what you're doing and what's not working.

**Comment by codechriscode** [comment]
Sorry, right

```
boundingVolumeHierarchy.shapecast({
  intersectsBounds: (box, isLeaf, score, depth, nodeIndex) => {
    accumulatedTraversedNodeIndices.add(nodeIndex);
    const intersects = sphere.intersectsBox(box);
    const { min, max } = box;
    if (intersects) {
      for (let x = 0; x <= 1; x++)
        for (let y = 0; y <= 1; y++)
          for (let z = 0; z <= 1; z++) {
            temporaryVector.set(
              x === 0 ? min.x : max.x,
              y === 0 ? min.y : max.y,
              z === 0 ? min.z : max.z
            );
            if (!sphere.containsPoint(temporaryVector)) return INTERSECTED;
          }
      return CONTAINED;
    }
    return NOT_INTERSECTED;
  },
```

The irony is that most OBJ-based geometries work fine with this. A few specific ones keep reproducing the error by never returning INTERSECTED, since the sphere never `.containsPoint`, because the tempVector (with box's min or max) is always from a depth 1 box, which is always too large to be contained in the sphere, and so on...

**Comment by gkjohnson** [comment]
I'll still need a working example of the issue, I think. I don't see anything in the provided code that looks immediately incorrect.

To this question

> The meshes result of OBJ files. MeshBVHVisualizer shows BVs as expected, but I'm not sure the hierarchy tree is correct since depth1 BVs are not sending me forward to leaf nodes. Is there a way to visualize the hierarchy to see if the tree data structure is well formed?

From the docs - the visualizer lets you display the parent bounds and control the displayed depth. There's also an undocumented [validateBounds](https://github.com/gkjohnson/three-mesh-bvh/blob/f4ea4350d97fdd9b1db459fd5bdae4241698c059/src/debug/Debug.js#L165) function which will check to ensure the child bounds are correctly encapsulated by the parent bounds. This was needed due to floating point error issues when storing parent bounds.

**Comment by codechriscode** [comment]
The tree is indeed well formed and passed validateBounds every time.

The issue was that the sphere was lacking a `sphere.applyMatrix(invMatrix)` to have the correct point coordinates in its own space.

Thanks for your support. I'm untagging "bug" and closing the issue.

**Comment by gkjohnson** [comment]
Ah yes - everything must be in the local frame of the BVH to work
